File: androidApp/src/main/kotlin/it/mensa/app/features/onboarding/OnboardingViewModel.kt
package it.mensa.app.features.onboarding

import androidx.lifecycle.ViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update

data class OnboardingPage(
    val icon: String,        // Material icon name reference for placeholder
    val title: String,
    val subtitle: String,
)

data class OnboardingUiState(
    val currentPage: Int = 0,
    val pages: List<OnboardingPage>,
) {
    val isLastPage: Boolean get() = currentPage == pages.size - 1
    val totalPages: Int get() = pages.size
}

class OnboardingViewModel(
    private val onComplete: () -> Unit,
) : ViewModel() {

    // Pages are populated at VM construction time so the i18n tr() calls
    // happen when the catalog is already loaded (bootstrapped before this phase).
    private val _pages: List<OnboardingPage> = buildPages()

    private val _uiState = MutableStateFlow(OnboardingUiState(pages = _pages))
    val uiState: StateFlow<OnboardingUiState> = _uiState.asStateFlow()

    fun onNext() {
        _uiState.update { state ->
            if (state.currentPage < state.pages.size - 1) {
                state.copy(currentPage = state.currentPage + 1)
            } else state
        }
    }

    fun onPageSelected(index: Int) {
        _uiState.update { it.copy(currentPage = index.coerceIn(0, it.pages.size - 1)) }
    }

    fun onComplete() {
        onComplete.invoke()
    }

    private fun buildPages(): List<OnboardingPage> = listOf(
        OnboardingPage(
            icon = "sparkles",
            title = "Benvenuto in Mensa",
            subtitle = "Sei entrato a far parte del 2% più brillante. Ti aiutiamo a sfruttare al meglio la community.",
        ),
        OnboardingPage(
            icon = "calendar_add_on",
            title = "Eventi in tutta Italia",
            subtitle = "Trova cene, conferenze e incontri vicino a te — o quelli nazionali. Iscriviti con un tap e aggiungili al calendario.",
        ),
        OnboardingPage(
            icon = "badge",
            title = "La tua tessera, sempre con te",
            subtitle = "Mostrala al coordinatore quando partecipi a un evento. Niente più tessere di plastica da cercare nel portafoglio.",
        ),
        OnboardingPage(
            icon = "search",
            title = "Una ricerca per tutto",
            subtitle = "Persone, eventi, convenzioni, documenti — un solo posto per trovare quello che ti serve.",
        ),
    )
}

File: iosApp/iosApp/Features/Onboarding/OnboardingViewModel.swift
import Foundation
import Shared

struct OnboardingPage {
    let icon: String
    let title: String
    let subtitle: String
}

@MainActor @Observable
final class OnboardingViewModel {
    var currentPage: Int = 0

    // Strings are resolved at VM construction via `tr(...)` literal calls so
    // the tolgee-push extractor picks up the keys + fallbacks.
    let pages: [OnboardingPage] = [
        OnboardingPage(
            icon: "sparkles",
            title: tr("onboarding.welcome.title", fallback: "Benvenuto in Mensa"),
            subtitle: tr("onboarding.welcome.subtitle", fallback: "Sei entrato a far parte del 2% più brillante. Ti aiutiamo a sfruttare al meglio la community.")
        ),
        OnboardingPage(
            icon: "calendar.badge.plus",
            title: tr("onboarding.events.title", fallback: "Eventi in tutta Italia"),
            subtitle: tr("onboarding.events.subtitle", fallback: "Trova cene, conferenze e incontri vicino a te, anche quelli nazionali. Iscriviti con un tap e aggiungili al calendario.")
        ),
        OnboardingPage(
            icon: "person.text.rectangle.fill",
            title: tr("onboarding.card.title", fallback: "La tua tessera, sempre con te"),
            subtitle: tr("onboarding.card.subtitle", fallback: "Mostrala al coordinatore quando partecipi a un evento. Niente più tessere di plastica da cercare nel portafoglio.")
        ),
        OnboardingPage(
            icon: "magnifyingglass",
            title: tr("onboarding.search.title", fallback: "Una ricerca per tutto"),
            subtitle: tr("onboarding.search.subtitle", fallback: "Persone, eventi, convenzioni, documenti. Un solo posto per trovare quello che ti serve.")
        )
    ]

    var isLastPage: Bool { currentPage == pages.count - 1 }

    private let onComplete: () -> Void

    init(initialPage: Int = 0, onComplete: @escaping () -> Void) {
        self.currentPage = initialPage
        self.onComplete = onComplete
    }

    func next() {
        guard currentPage < pages.count - 1 else { return }
        currentPage += 1
    }

    func complete() {
        // In DEBUG we deliberately DON'T persist the "shown" marker, so every
        // future login re-triggers the onboarding (paired with the
        // evaluatePhase() #if DEBUG branch in RootViewModel). The user still
        // enters the app normally on tap.
        #if !DEBUG
        if let uid = koin.auth.currentUser.value as? UserModel {
            koin.onboarding.markShown(userId: uid.id)
        }
        #endif
        onComplete()
    }
}
